File: lintcode/easy/235_prime_factorization.py
'''
Link: https://www.lintcode.com/problem/prime-factorization/description
'''
# This solution is inspired by the explanation on Jiuzhang.com
class Solution:
    """
    @param num: An integer
    @return: an integer array
    """
    def primeFactorization(self, num):
        # write your code here
        prime_list = []
        if num is None or num <= 0:
            return prime_list
        ceiling = int(num ** 0.5) + 1
        for denom in range(2, ceiling):
            while num % denom == 0:
                num = num/denom
                prime_list.append(denom)
            if num == 1:
                return prime_list
        prime_list.append(int(num))
        return prime_list

File: lintcode/easy/235_prime_factorization.java
/* This script is written based on the solution in Jiuzhang.com, they are very similar.
*/

public class Solution {
    /**
     * @param num: An integer
     * @return: an integer array
     */
    public List<Integer> primeFactorization(int num) {
        // write your code here
        List<Integer> results = new ArrayList<>();
        int max = (int) Math.sqrt(num);
        for (int i = 2; i <= max && num > 1; i++){
            while (num % i == 0){
                num = num / i;
                results.add(i);
            }
        }
        if (num > 1){
            results.add(num);
        }
        return results;
    }
}
